rpcs3/Emu/RSX/RSXFIFO.cpp

```diff
@@ -17,26 +17,44 @@ namespace rsx
 		{
 			m_internal_get += 4;
 
-			if (wait && m_ctrl->put == m_internal_get)
+			if (wait && read_put() == m_internal_get)
 			{
 				// NOTE: Only supposed to be invoked to wait for a single arg on command[0] (4 bytes)
 				// Wait for put to allow us to procceed execution
 				sync_get();
 
-				while (m_ctrl->put == m_internal_get && !Emu.IsStopped())
+				while (read_put() == m_internal_get && !Emu.IsStopped())
 				{
 					std::this_thread::yield();
 				}
 			}
 		}
 
-		void FIFO_control::set_put(u32 put)
+		u32 FIFO_control::read_put()
 		{
-			if (m_ctrl->put == put)
+			// Lowest 2 bits, as BE value
+			constexpr u32 align_bits = 3 << (IS_LE_MACHINE * 24);
+
+			u32 result;
+			m_ctrl->put.fetch_op([&](be_t<u32>& val)
```

Comment:
I'm fairly certain this need not be atomic. Race conditions between RSX and Cell play out naturally; I think just returning put.raw() & align is enough for this.

Reply:
it doesnt really matter for performance if the clearing bits op is atomic or not, because the operation is caneclled if it is already aligned. See fetch_op implementation.

Reply:
Are the branches still faster than InterlockedAnd though? Least loc would just be raw & or even var.and_fetch(). I guess it all depends on what the compilers generate for the op inlining.

Reply:
MSVC generates 3 more bswaps with and_fetch(). R/W operations with byeswapping are tricky.

Reply:
Fair enough. I get uneasy if too many ops are needed in the core 'tick' logic since it tends to slowly eat away at performance over many commits.
May be worth specializing be_t logical operations? Logically you only need one byteswap for that if the compiler was good. I'll play around with msvc after merge and see if it can work with something nicer.

Reply:
This raw() hand "optimizations" are like plague in the code. Need to stop it before it spreads.

Reply:
"ok"